Audio: Add Sample Rate as a settable option in the Operator Menu

pnn64 opened the pull request:
### TL;DR
This PR changes StepMania's default internal/output sample rate from 44.1kHz to 48kHz, aligning with modern hardware and reducing system-level resampling that introduces sync drift.

### Background / Story
Sync issues have haunted StepMania and its forks since the 3.95 era—mid-song drift, inconsistent start sync, break-time desync, and even progressive drift after idling. Frustration within the community—especially highlighted during events like the recent ITL tournament—has underscored the critical need for a solution.

Historically, StepMania has defaulted to an internal audio processing and output sample rate of 44.1kHz, a legacy standard from the CD audio era. However, modern operating systems and audio hardware now predominantly use 48kHz (or higher, e.g., 96kHz). When StepMania outputs audio at 44.1kHz to a system expecting 48kHz, the OS or audio driver performs a real-time resampling step. This external resampling is outside of StepMania's control and can introduce non-deterministic latency and cumulative drift, which are the root causes of many sync problems observed in gameplay.

The key insight for this PR emerged from a three-week deep dive into building a Vulkan-based StepMania engine in Rust—[DeadSync](https://github.com/pnn64/deadsync). Frustrated by StepMania's lingering sync issues, I built DeadSync as an experimental engine designed to solve them from the ground up.

While implementing song previews (which involves seeking into long .ogg files), I began exploring how sample rates affect playback. DeadSync, too, despite meticulous timing logic, exhibited severe sync drift—even worse than StepMania—when using Rust audio libraries that resampled 44.1kHz source files to a 48kHz output.

The breakthrough came when I manually converted an .ogg file to 48kHz and configured the engine to output at 48kHz as well. With all resampling layers eliminated, the sync was immediately stable. On the first attempt, I nearly quadded a 15-footer (on keyboard) without any global offset calibration.

I shared this finding with @sukibaby (jasmine), who quickly confirmed that StepMania's audio code defaults internally to 44.1kHz. She created a test build forcing 48kHz internally, and user Kigha reported immediate improvements in sync stability, stating:
> It has genuinely resolved everything and restored my skepticism of drift sync while playing.

### This PR
This pull request addresses the core issue by changing StepMania's default internal and output sample rate from 44.1kHz to 48kHz. This change aligns StepMania's audio pipeline with modern audio hardware defaults and reduces or eliminates the need for system-level resampling.

### Next Steps
To further improve sync, I'll provide a script to convert .ogg and .mp3 song files to 48kHz. (This can also be done manually using Audacity or similar tools.) By matching the audio files to the system's native rate, we can minimize or eliminate resampling both in ITGmania and at the system/driver level—potentially improving not just sync stability, but the overall tightness of the timing.

For now, this PR focuses solely on changing the engine’s default sample rate. I'll share the conversion script in the sync forum on Discord, starting with all ITL charts.

### Acknowledgement
Thanks to @sukibaby for letting me do this PR and for the feedback and encouragement. It's wild that after so many years, something this simple might be a major step toward finally fixing rhythm gaming’s longest-running sync mystery.

pnn64 committed:
Audio: Update default sample rate to 48kHz to mitigate sync issues

PawelA commented:
That's a great find!
Though wouldn't that make the situation worse for people with 44.1kHz sound cards? (I'm pretty sure my laptop has one, can't check rn) Some kind of check that chooses the right one would work better I think.
Even better, how much work would it be to target a low-level API that bypasses the OS-level resampling completely? For windows I don't know, but on linux it should definitely be possible with the ALSA backend.

pnn64 commented:
You've hit on a key point—blindly switching to 48kHz without an override could absolutely cause issues for users on hardware that still natively runs at 44.1kHz.

That said, the vast majority of modern systems (especially on Windows) default to 48kHz, and this PR is aimed at improving the experience for that majority—where resampling from 44.1kHz introduces unpredictable drift. It’s likely that users on true 44.1kHz hardware weren’t affected by this particular issue in the first place.

But you're totally right: the ideal solution is to make this configurable. Something like `SoundSampleRate=Auto|44100|48000` in `Preferences.ini`, with `Auto` doing native sample rate detection and falling back to a sane default. That would ensure maximum compatibility. Dynamic detection of the native sound card sample rate should not be too hard to implement.

teejusb commented:
I think defaulting to 48kHz for everything will cause some regressions depending on the OS/hardware setups. On my Windows/Mac setup it does seem like my devices are 48kHz by default, but for my Linux setup it's seems to be 44,100Hz so this will likely regress those setups.

```sh
[21:02:45] itg@itg:~$ pactl list short sinks
1	alsa_output.usb-Generic_Blue_Microphones_2050BAB01U28-00.analog-stereo	module-alsa-card.c	s16le 2ch 44100Hz	SUSPENDED
2	alsa_output.pci-0000_00_1f.3.analog-stereo	module-alsa-card.c	s16le 2ch 44100Hz	RUNNING
12	alsa_output.pci-0000_01_00.1.hdmi-stereo	module-alsa-card.c	s16le 2ch 44100Hz	SUSPENDED
13	nx_voice_out	module-null-sink.c	s16le 2ch 44100Hz	IDLE
```

```sh
[21:01:01] itg@itg:~$ arecord -D hw:0 --dump-hw-params
HW Params of device "hw:0":
--------------------
ACCESS:  MMAP_INTERLEAVED RW_INTERLEAVED
FORMAT:  S16_LE S32_LE
SUBFORMAT:  STD
SAMPLE_BITS: [16 32]
FRAME_BITS: [32 64]
CHANNELS: 2
RATE: [44100 192000]
PERIOD_TIME: (83 11888617)
PERIOD_SIZE: [16 524288]
PERIOD_BYTES: [128 2097152]
PERIODS: [2 32]
BUFFER_TIME: (166 23777234)
BUFFER_SIZE: [32 1048576]
BUFFER_BYTES: [128 4194304]
TICK_TIME: ALL
--------------------
```

Just for curiosity, I also looked at the distribution of sample rates of my ITG library. It is predominantly 44,100, which makes sense since it's standard for digital audio to be 44,100.
```
44100: 12942
48000: 1531
44099: 66
32000: 17
22050: 14
96000: 10
11025: 2
24000: 2
86000: 2
136100: 2
100000: 1
113000: 1
126000: 1
192000: 1
```

I think that indicates that we'd be resampling somewhere along the way. Either from ITGmania itself, or when interfacing with the device itself.

I think the better general solution is that if `m_iSoundPreferredSampleRate` is 0 (which is the default) then we probe the sound device for its sample rate instead of relying on the hardcoding values you're updating in this PR. That way we wouldn't regress *any* setup as we'd always be matching the hardware.

Along that line, have we tested if explicitly setting 44,100 in sound settings helps users on release builds? I'm curious about those results. Similarly curious about results where we set the preference to match the audio device (i.e. 48,000 for most setups).

![image](https://github.com/user-attachments/assets/7196b547-3011-4650-986b-45ccd507f49e)

Fwiw, I'm not entirely opposed to defaulting to 48kHz (and is actually generally preferred), but I'm trying to assess the best course of action here. This is still such great insight so thanks for this!

(I recognize this is similar to pnn64's comment above, but I do think it's very worthwhile to explore the option instead of necessarily updating the hardcoded constant).

sukibaby commented:
> I think defaulting to 48kHz for everything will cause some regressions depending on the OS. On my Windows/Mac setup it does seem like my devices are 48kHz by default, but for my Linux setup it's seems to be 44,100Hz so this will likely regress those setups.

Just quickly mentioning on this point,

We did find it's default 44.1khz for Mac users, so I know we do need to account for the user's OS settings.

I'm currently writing a method to get the default audio device's bit depth and sample rate so that we can match the OS before spinning up an audio decoding thread.

jsirex commented:
It looks like on Linux (Debian Testing) with pipewire I have 48K
```
❯ grep -B 4 -A 4 "default.clock.rate" /usr/share/pipewire/pipewire.conf
    core.daemon = true              # listening for socket connections
    core.name   = pipewire-0        # core name and socket name

    ## Properties for the DSP configuration.
    #default.clock.rate          = 48000
    #default.clock.allowed-rates = [ 48000 ]
    #default.clock.quantum       = 1024
    #default.clock.min-quantum   = 32
    #default.clock.max-quantum   = 2048
```

`pw-cli info all | grep clock.rate` returns same

teejusb commented:
I think this just reinforces that assuming a default sample rate is likely not the strategy consider the wide array of hardware.

teejusb commented:
A simple interim solution (if we don't figure out auto sample rate detection sooner than later) is to add preferred sample rate as an option in the operator menu. That way we could define Default, 44100, 48000 etc as selectable options and have users decide.

Actually I think adding this is likely good in general, and then "Auto" could later be added as an option (and then made the default once it works). I think I'm leaning towards that solution vs updating the default in this PR as this seems more disruptive to current behavior (basically making good setups bad, and bad setups good in a somewhat opaque way).

pnn64 committed:
Implement configurable Sample Rate in options

Key changes:
- Added "Sample Rate" to `ScreenOptionsGraphicsSound` with choices:
    - "Default" (resolves to 44100 Hz for now)
    - "44100 Hz"
    - "48000 Hz"
- Created a new ConfOption "PreferredSampleRate" in `ScreenOptionsMasterPrefs.cpp` mapping UI choices to the integer values (0, 44100, 48000) for the existing `m_iSoundPreferredSampleRate` preference.
- Updated relevant sound drivers:
    - DirectSound (RageSoundDriver_DSound_Software): Primary buffer now attempts to use the preferred rate. Secondary buffers already handled it correctly.
    - WaveOut (RageSoundDriver_WaveOut): Already handled 0 as 44.1kHz correctly.
    - PulseAudio (RageSoundDriver_PulseAudio): Ensured m_InitStream uses the constructor-initialized rate based on preference.
    - AudioUnit (RageSoundDriver_AU): Already handled 0 as 44.1kHz and attempts to set hardware rate.
    - OSS (RageSoundDriver_OSS): Updated to read and apply the preferred sample rate.
    - WDMKS (RageSoundDriver_WDMKS): Updated to pass the resolved preferred rate to its stream opening logic.
    - Null (RageSoundDriver_Null): Already handled 0 as 44.1kHz correctly.
- Updated `RageSoundReader_Chain` constructor to use the preferred system sample rate for its internal default.
- Added English localization for the new option title and explanation.

The implementation ensures that if the preference is 0 (Default), the system currently defaults to 44100 Hz across all drivers. This lays the groundwork for future dynamic hardware rate detection. A game restart is recommended for the new sample rate to take full effect.

sukibaby commented on src/RageSound.cpp:
```diff
@@ -147,7 +147,7 @@ class RageSoundReader_Silence: public RageSoundReader
 	int Read( float *pBuf, int iFrames ) { return RageSoundReader::END_OF_FILE; }
 	RageSoundReader *Copy() const { return new RageSoundReader_Silence; }
-	int GetSampleRate() const { return 44100; }
+	int GetSampleRate() const { return 44100; } // Hardcoded to 44100
```
might recommend making this value a constant  (`constexpr int kFallbackSampleRate = 44100` or something) so that it can easily be changed in the future anywhere that's relevant if it's needed - i think this could be defined in RageSound.h so that any thing RageSound* can use it

pnn64 replied:
Absolutely, something like `constexpr int K_DEFAULT_FALLBACK_SAMPLE_RATE = 44100;`

defined in `RageSound.h`. As you pointed out, this will make it much clearer in the codebase when `44100` is being used as an intentional fallback versus an actively configured or hardware-derived rate.

sukibaby replied:
`K_DEFAULT_FALLBACK_SAMPLE_RATE` is macro case,  we'd want it to be stylized like `kFallbackSampleRate`

sm5/ITGm does still have some lingering constants as macro case but i've been changing them over as i've changed them from `#define`s to `constexpr`s.

sukibaby commented on Themes/_fallback/Languages/en.ini:
```diff
@@ -1203,6 +1204,7 @@ Player1Profile=Player1 Profile
 Player2Profile=Player2 Profile
 PlayerAutoPlay=AutoPlay
 Predicted Meter=Predicted Meter
+PreferredSampleRate=Sample Rate
```
personally i might think "Audio Samplerate" or something so it's more obvious to people that this setting corresponds to sound stuff?

pnn64 replied:
Agreed, maybe something like: PreferredSampleRate=Audio Sample Rate

sukibaby reviewed (commented):
I read it over, looks fine. Just a few nitpicks in this review, tbh.

pnn64 commented:
This latest commit implements the initial step of @teejusb's proposed plan to address audio sample rate handling. Based on our discussion, we're adding an explicit "Sample Rate" option in the settings with choices for "Default" (currently 44.1kHz), "44100 Hz," and "48000 Hz."

The primary goal is to better manage the chain: sample rate of audio file -> preferred sample rate of game -> sample rate of hardware. This change focuses on allowing users to control the "preferred sample rate of game." The motivation is that sync issues often appear when the game's output sample rate doesn't align well with the OS/hardware, particularly on Windows systems. By letting the game output at a well-defined rate (and handle internal resampling to that rate), we aim to reduce reliance on potentially problematic OS-level resampling.

Apologies if there are any C++ faults in the implementation; it's been over a decade since I've worked with the language, so any feedback is greatly appreciated!

pnn64 committed:
Add "Audio Sample Rate" option with translations

The English display name for the option has been updated to "Audio Sample Rate" for better clarity. Translations for this new option title have been added for Spanish, French, German, Japanese, Korean, Dutch, Polish, Slovak, and Traditional Chinese.

pnn64 committed:
Centralize fallback sample rate with kFallbackSampleRate

Replaces hardcoded 44100Hz values (used as fallbacks or for the "Default" sample rate preference) with a new constant `constexpr int kFallbackSampleRate = 44100;` defined in RageSound.h.

This constant is now used in:
- RageSoundReader_Silence
- RageSoundManager (for unloaded driver scenarios)
- All sound drivers' initialization logic for the "Default (0)" preference (DSound, WaveOut, AU, PulseAudio, OSS, WDMKS, Null)
- DSound primary buffer setup.

pnn64 commented:
I've just committed two updates. The initial change provides a new name to the "Audio Sample Rate" option and includes the necessary localizations. The subsequent commit focuses on maintainability by implementing the `kFallbackSampleRate` constant, which makes the code's intent clearer when dealing with default audio rates. The build compiled successfully, and I'll be testing the functionality later today.

teejusb commented on src/RageSoundManager.cpp:
```diff
@@ -133,9 +133,9 @@ float RageSoundManager::GetPlayLatency() const
+		return kFallbackSampleRate;
 
-	return m_pDriver->GetSampleRate();
+	return m_pDriver->GetSampleRate(); // Returns the *actual* operating rate of the loaded driver
```
Here and generally throughout the PR, let's put the comment above instead of on the same line (since it'll be safer for the line length limit)

teejusb commented on src/RageSoundReader_Chain.cpp:
```diff
@@ -23,7 +23,10 @@
-	m_iPreferredSampleRate = 44100;
+	m_iPreferredSampleRate = PREFSMAN->m_iSoundPreferredSampleRate;
+	if (m_iPreferredSampleRate == 0)
+		m_iPreferredSampleRate = 44100; // Default if system default (0) is chosen
```
Add braces.

Also The comment here is unnecessary since it's obvious from the code.

Also Also, any reason we're not using the constant here?

pnn64 committed:
Address review feedback and improve code style

- Moved end-of-line comments to preceding lines for better readability and to avoid potential line length issues across various files modified in previous commits.
- In the RageSoundReader_Chain constructor: Removed redundant initialization of m_iPreferredSampleRate. Replaced hardcoded 44100 with kFallbackSampleRate. Removed an unnecessary comment.
- Added braces consistently to single-statement if/else blocks across all recently modified sound system files. This enhances code clarity and maintainability by explicitly defining block scopes.

pnn64 commented:
Thanks for the review @teejusb. I've committed the latest changes based on your recommendations.

DolphinChips commented on src/RageSoundManager.cpp:
```diff
@@ -133,9 +133,10 @@ float RageSoundManager::GetPlayLatency() const
 
-	return m_pDriver->GetSampleRate();
+	// Returns the *actual* operating rate of the loaded driver
+	return m_pDriver->GetSampleRate(); 
```
Trailing whitespace on this line

pnn64 committed:
Remove one trailing whitespace
